src/coreclr/jit/importercalls.cpp

```diff
@@ -3987,6 +3988,49 @@ GenTree* Compiler::impSRCSUnsafeIntrinsic(NamedIntrinsic        intrinsic,
         {
             assert((sig->sigInst.methInstCount == 1) || (sig->sigInst.methInstCount == 2));
 
+            // The signature is:
+            //
+            //   T As<T>(object o) where T : class
+            //
+            // it means we need to query the exact T since it's always a shared type due
+            // to its constraint.
+            //
+            if (sig->sigInst.methInstCount == 1)
+            {
+                CORINFO_GENERICHANDLE_RESULT embedInfo;
+                info.compCompHnd->embedGenericHandle(pResolvedToken, false, &embedInfo);
+                if (!embedInfo.lookup.lookupKind.needsRuntimeLookup)
```

Comment:
I think you want to check for `CORINFO_FLG_SHAREDINST` flag here. 

I do not think you need to re-resolve the signature when it is not shared. `sig->sigInst.methInst[0]` should have the exact type for the non-shared case.

Reply:
> I do not think you need to re-resolve the signature when it is not shared. sig->sigInst.methInst[0] should have the exact type for the non-shared case.

Thanks, I was wondering about that too. I though that since I only need it here I can request that on demand, will change the original sig.

Reply:
@jkotas it seems that we're losing the exact sig due to this line of `getCallInfo`: https://github.com/dotnet/runtime/blob/231314d2ff71e6c8e8c9d2e9c7e6a8de91224cd2/src/coreclr/vm/jitinterface.cpp#L5232

this wrapped pMD is then used for `getMethodSigInternal` does it look correct to you?

Reply:
Myabe it should be here `if (pTargetMD->IsWrapperStub())` ?

Reply:
Does it work with the last update, or do you still need help with this?

Reply:
> Does it work with the last update, or do you still need help with this?

Still need help, the current impl doesn't work since we always have System._Canon as methodInst[0], I just prepared the code to pick up a fix in the getCallInfo

Reply:
Ah I see, thanks! I assume we only need to do that for `Unsafe.As`?